### PR TITLE
Add regression test for #86106

### LINKED ISSUE #86106: `String::new()` is slower than `"".to_string()`
I tried this code:

```rust
pub fn a() -> String {
    String::new()
}

pub fn b() -> String {
    "".to_string()
}
```

I expected to see this happen: both functions contain identical code.

Instead, this happened: https://rust.godbolt.org/z/P53MMG4W6.

For `String::new()` it does one more instruction: `mov     rcx, qword ptr [rip + .L__unnamed_1]`. This `rcx` is then read further below at `mov     qword ptr [rdi], rcx`.
The `"".to_string()` version however does not do this extra step and immediately reads from a constant: `mov     qword ptr [rdi], 1`.
This means that `String::new()` does slightly more work than `"".to_string()` but I expect them to be identical.

Before 1.52.0 the functions generated identical code: https://rust.godbolt.org/z/8647E6YhY.

oli noted the following:
> The problem is probably because it copies its value from the constant https://github.com/rust-lang/rust/blob/d620ae10709ca3669cda89db8b29afcd9accc188/library/alloc/src/raw_vec.rs#L64 instead of immediately creating the appropriate value

### Meta

This currently happens on nightly.

### PR DESCRIPTION
**Pull request by JohnTitor**
Closes #86106
r? @nikic 

Signed-off-by: Yuki Okushi <jtitor@2k36.org>

### CODE REVIEW

File: src/test/codegen/issue-86106.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// min-llvm-version: 15.0
+// compile-flags: -C opt-level=3 -C target-cpu=native
```
Comment (nikic) [inline src/test/codegen/issue-86106.rs:2]: I don't think we need `target-cpu=native`.

File: src/test/codegen/issue-86106.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+// The below two functions ensure that both `String::new()` and `"".to_string()`
+// produce the identical code.
+
+#![crate_type = "lib"]
+
+// CHECK-LABEL: @string_new = unnamed_addr alias void (ptr), ptr @empty_to_string
+// CHECK: define void @empty_to_string
```
Comment (nikic) [inline src/test/codegen/issue-86106.rs:10]: This should be CHECK-LABEL and closer to the empty_to_string function.

File: src/test/codegen/issue-86106.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+#[no_mangle]
+pub fn empty_to_string() -> String {
+    // CHECK-NOT: load i8
+    // CHECK: store i64
+    // CHECK-NEXT: getelementptr
+    // CHECK-NEXT: store ptr
+    // CHECK-NEXT: getelementptr
+    // CHECK-NOT: store i8
```
Comment (nikic) [inline src/test/codegen/issue-86106.rs:23]: This CHECK-NOT is redundant when using CHECK-NEXT.

File: src/test/codegen/issue-86106.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// CHECK: define void @empty_to_string
+#[no_mangle]
+pub fn string_new() -> String {
+    String::new()
+}
+
+#[no_mangle]
+pub fn empty_to_string() -> String {
+    // CHECK-NOT: load i8
+    // CHECK: store i64
```
Comment (nikic) [inline src/test/codegen/issue-86106.rs:19]: I think this test is going to fail on 32-bit targets, where usize is 32-bit.

File: src/test/codegen/issue-86106.rs (hunk 1 above)
Comment (JohnTitor) [reply src/test/codegen/issue-86106.rs]: Thank you! Fixed with other comments via https://github.com/rust-lang/rust/pull/106036/commits/3b16aea1237617578451d947a87f057fcc1250b3.

**Comment by JohnTitor** [comment]
Is there a way to avoid running the test on nopt or force optimization?

**Comment by nikic** [comment]
`-C opt-level=3` already forces optimization. From the output, it's not completely clear why the test fails. I've submitted https://github.com/rust-lang/rust/pull/106078 to make this easier to debug in the future.

Possibly it's just a function merging failure. You could try passing `-Z merge-functions=disabled` and test the result of that (without the alias).

**Comment by JohnTitor** [comment]
Sounds good, applied it: https://github.com/rust-lang/rust/pull/106036/commits/ab9313e624d581a75f4b886c0adf351f848d725a

**Pull request merged**